### PR TITLE
Add group by functionality

### PR DESCRIPTION
**Pull request by wrandall22**

### CODE REVIEW

**Comment by wrandall22** [comment]
I should still add `Having` functionality to this.

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/AbstractDao.java [hunk 1]
```diff
@@ -216,9 +216,10 @@ public final Cursor getCursor(@NonNull final Class<?> clazz, @NonNull final Stri
     @WorkerThread
     public final <T> Cursor getCursor(@NonNull final Class<T> clazz, @NonNull final Join<T, ?>[] joins,
                                       @NonNull final String[] projection, @Nullable final String whereClause,
-                                      @Nullable final String[] whereArgs, @Nullable String orderBy) {
+                                      @Nullable final String[] whereArgs, @Nullable String orderBy,
+                                      @Nullable final String groupBy) {
```
Comment (frett) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/AbstractDao.java]: let's not add groupBy to this method, it's deprecated and kept around for compatibility with projects I haven't converted to use the Query object yet.

**Comment by frett** [comment]
+1 unless you were going to add the `having` functionality, which should emulate the where functionality

**Comment by wrandall22** [comment]
I should probably add it before I merge it, but I'll wait until Monday unless you need it now.

**Comment by frett** [comment]
nope, I'm not waiting on it.

I'm currently working up some unit tests for the eventbus loader helper. There are quite a few complexities & use cases involved

**Comment by frett** [comment]
I fixed the crash, but may have introduced a different issue. so I really want to setup a test to ensure it behaves correctly.

**Comment by wrandall22** [comment]
Updated the code for this to support having, and fix a couple bugs.

**Comment by frett** [comment]
If I remember correctly, having is more like a where clause than like group by or order by. So, the having field should probably be an expression and go through the SQL generation process

**Comment by wrandall22** [comment]
True that. Having is a where clause for group by statements.

**Comment by frett** [comment]
also, since this is new and doesn't need the backwards compatibility, let's making mGroupBy a `Field[]` which encompasses the whole prefix functionality.

**Comment by frett** [comment]
so, `Query.select(Table.class).groupBy(FIELD_A, FIELD_B)` using varargs

**Comment by wrandall22** [comment]
One thing about the having clause is that the SQLite stuff doesn't have args for it, so it has to be put in as a string. So wherever that's abstracted to, needs to be on our side, so I'll do some work to turn expression back to string.

**Comment by frett** [comment]
Take a look at how I handle joins, same thing applies there

**Comment by wrandall22** [comment]
Rebased with master, made some changes, added some tests.

File: gto-support-db/src/androidTest/java/org/ccci/gto/android/common/db/AbstractDaoTest.java [hunk 2]
```diff
@@ -22,4 +24,102 @@ public void testGetTable() throws Exception {
         final TestDao dao = getDao();
         assertThat(dao.getTable(Root.class), is(RootTable.TABLE_NAME));
     }
+
+    @Test
+    public void testGetCursor() {
+        final TestDao dao = getDao();
+        final Cursor cursor = dao.getCursor(Query.select(Root.class));
+
+        assertThat(cursor.getColumnName(0), is("_id"));
```
Comment (frett) [inline gto-support-db/src/androidTest/java/org/ccci/gto/android/common/db/AbstractDaoTest.java]: I would change this test to ensure that the default columns are present, not in a specific position.
so, something like:
`assertThat(cursor.getColumnIndex(COLUMN_ID), is(not(-1)))`

Comment (frett) [inline gto-support-db/src/androidTest/java/org/ccci/gto/android/common/db/AbstractDaoTest.java]: that way ordering of columns within the Cursor isn't being tested, but column presence is

File: gto-support-db/src/androidTest/java/org/ccci/gto/android/common/db/ExpressionTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        Expression equalsExpression = RootTable.FIELD_TEST.eq("1");
+        assertThat(equalsExpression.buildSql(dao).first, is("(root.test == ?)"));
+    }
+
+    @Test
+    public void testNotEqualsSql() {
+        final TestDao dao = getDao();
+
+        Expression notEqualsExpression = RootTable.FIELD_TEST.not().eq("1");
```
Comment (frett) [inline gto-support-db/src/androidTest/java/org/ccci/gto/android/common/db/ExpressionTest.java]: so, while this test is valid, this isn't something that should probably happen in normal usage. the expression you are probably looking for is: `FIELD_TEST.eq("1").not()` or `FIELD_TEST.ne("1")`

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java [hunk 4]
```diff
@@ -760,4 +800,89 @@ public Unary createFromParcel(@NonNull final Parcel in, @Nullable final ClassLoa
...
+    public static class Aggregate extends Expression {
+        static final String COUNT = "COUNT";
+        static final String MAX = "MAX";
+        static final String MIN = "MIN";
+        static final String SUM = "SUM";
+
+        @NonNull
+        private final String mOp;
+        @NonNull
+        private final Expression mExpr;
```
Comment (frett) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java]: I would also add:
```
@NonNull
private final boolean mDistinct;
```

Comment (frett) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java]: I would also narrow this to `private final Field mExpr` since Aggregate functions can only apply to fields (I believe). we can widen this down the line if other use cases actually show up.

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java [hunk 5]
```diff
@@ -133,6 +133,46 @@ protected Binary binaryExpr(@NonNull final String op, @NonNull final Expression
     }
 
     @NonNull
+    public Aggregate count() {
```
Comment (frett) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java]: these aggregate functions should be moved to the `Field` class per the previous comment of narrowing Aggregate scope to `Field`s only

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java (hunk 4 above)
Comment (wrandall22) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java]: Hmm, I was thinking this `mExpr` was to be used for queries like

````
SELECT ...
FROM ...
GROUP BY ...
HAVING count(select ... from ... where ...) > 1;
````

If this is not going to be useful, I can also get rid of the methods taking in parameters.

Comment (wrandall22) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java]: Though, I suppose that would require a `Query` not an `Expression`.

Comment (frett) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java]: Hmm, I think for now let's narrow the scope for our current use cases. We can figure out subselects in the future if the need arises, but expressions may not be correct for those, and we may want to keep subselects separate from aggregate functions

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/AbstractDao.java [hunk 6]
```diff
@@ -236,10 +239,9 @@ public final Cursor getCursor(@NonNull final Class<?> clazz, @NonNull final Stri
                 projection[i] = projection[i].contains(".") ? projection[i] : prefix + projection[i];
             }
 
-            // prefix an un-prefixed orderBy field
-            if (orderBy != null && !orderBy.contains(".")) {
-                orderBy = prefix + orderBy;
-            }
+            // prefix un-prefixed clauses
+            addPrefixToFields(orderBy, prefix);
```
Comment (frett) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/AbstractDao.java]: you lost the test to see if orderBy already has a prefix, this will end up breaking the measurements app, I know I utilize an order by with a prefix. This also lost the test for null orderBy.

Comment (wrandall22) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/AbstractDao.java]: Oops, yep, I did indeed.

Comment (frett) [inline gto-support-db/src/main/java/org/ccci/gto/android/common/db/AbstractDao.java]: and you aren't even capturing the response, let me rework how groupBy is handled, I'd like to use the Field object for those and avoid having to manually prefix strings.

**Comment by frett** [comment]
I made a couple adjustments, waiting for travis to finish, once that's done I'm +1 on this

**Comment by wrandall22** [comment]
Rebased master, used new deleteAll

**Comment by frett** [comment]
+1

**Comment by wrandall22** [comment]
Thanks!
